### DISCUSSION TITLE
Raw AMZ SDK works, but this library does not. Cannot authorize no matter what I do

### DISCUSSION THREAD

**Discussion by booktrakker** [abuzuhri/Amazon-SP-API-CSharp#205] [Q&A]
I am hoping to use this library for my development, but I cannot get authenticated.  I get various errors, but the primary one seems to be a complaint about using the ARN.

I can successfully authenticate and retrieve an order using the raw Amazon SDK, but not with this library.  Please respond soon, I am seeing messages without a response for over a week, and that greatly concerns me.  If questions do not get answered in a timely manner then the value of the library evaporates.

The Raw Amz SDK does not require an ARN, and I suspect this is a critical difference.  Why does this libary require one when the Raw SDK does not?

Whatever the answer, I need to get this working.

I hope there is a simple solution as I really need to get going.  

Thanks!

**Comment by abuzuhri** [comment]
What you mean with Raw Amazon ?

**Comment by booktrakker** [comment]
The actual Amazon SP-API C Sharp SDK that we have to build using Swagger tools.  I can make successful calls using their code, but not yours.

 

And thank you very much for getting back to me quickly, you vastly exceeded my concern in that regard and I am truly grateful.  Amazon has been…. Difficult.

 

I am the only developer and they seem to think I have a huge team and lots of experience with Java and the related tools.

 

Your library is so well designed from a usage perspective it will really accelerate my development, but I keep getting this error.  The user I am submitting does have the required permissions, at least, as I and the Amazon TAM I work with can tell.  So why do I keep getting this error?

 

User: arn:aws:iam::281028383786:user/BookTrakkerApp is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::281028383786:user/BookTrakkerApp

 



 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***

**Reply by abuzuhri** [reply]
I’m not sour why it’s work in AMAZON SDK and not work with us but I think you don’t add the roles in Amazon connection and you add user 

many developers use the library and it’s work with them including me .

try to read instructions and start again to build it .

for your information our library also base on Amazon code

**Reply by booktrakker** [reply]
Tareq, I have pored over the documentation and IAM settings and I cannot find any way past this issue.  I am stuck with no clue as to how to make my Role work.  I keep getting told it is not authorized, yet I cannot find any way to authorize it, and I have completely conformed to the Amazon Migration Guide.  I have read and re-read the instructions to no avail.  I cannot see anything wrong yet it fails every time.

**Comment by booktrakker** [comment]
Tareq,

 

Thanks for the quick response!

 

Yes, I am well aware you built the library on the Amazon code, that is a key feature.  I am very glad it is, as that makes it an extension and I may be able to get some insight from my Amazon TAM – although that is far from certain.

 

I actually am not clear on what you mean by this:

‘but I think you don’t add the roles in Amazon connection and you add user’

I have tried an IAM User and IAM role, neither work, and the error I get makes no sense as my user has the STS permissions.  This whole experience has been brutal.  I have been working on this for the better part of a year due to the lack of support from Amazon; it was only last week that I finally got them to understand I could not generate the Swagger libraries as I know nothing about Java and it just kept failing no matter what I did.

 

Now I am hung up with this ARN issue that I cannot resolve.  I have no idea why one method as shown below works while whatever is happening in your library fails.

 

All I know is that I fill out the Credentials and get that error, or others, no matter what I do, yet the code test that uses a very different credentialing strategy works when I test it in the Amazon code.

 

I sent what you see below to my TAM.  If you could review it and help me understand what is going on it would be fantastic.

 

 

Here is  how it is handled in the IO.Swagger.Test:

 

LWAAuthorizationCredentials lwaAuthorizationCredentials = new LWAAuthorizationCredentials

            {

                ClientId = "",

                ClientSecret = "",

                RefreshToken = "",

 

                Endpoint = new Uri(https://api.amazon.com/auth/o2/token)

            };

 

            AWSAuthenticationCredentials awsAuthenticationCredentials = new AWSAuthenticationCredentials

            {

                AccessKeyId = "", //Get these right and see what happens.

                SecretKey = "",

                Region = "us-east-1"

            };

 

            instance = new OrdersV0Api.Builder()

                .SetAWSAuthenticationCredentials(awsAuthenticationCredentials)

                .SetLWAAuthorizationCredentials(lwaAuthorizationCredentials)

                .Build();

 

Here is what I do in the Wrapper Library.  I am pretty certain tha the AmazonCredentials is something that came from the Amazon SDK itself.  I am trying to understand how this is used and it is not exactly clear.

 

mAmzConnection = new AmazonConnection(new AmazonCredential()

                    {

                        //AccessKey = cAwsUserAccessKey,

                        //SecretKey = cAwsUserSecretKey,

                        AccessKey = cSpApiClientAccessKey,

                        SecretKey = cSpApiSecretAccessKey,

                        RoleArn = cAmzArn,

                        ClientId = cSpApiClientId,

                        ClientSecret = cSpApiClientSecretKey,

                        RefreshToken = mRefreshToken,

                        MarketPlace = marketPlace //MarketPlace.GetMarketPlaceByID("A2VIGQ35RCS4UG") 

                    });

 

 

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***

**Comment by booktrakker** [comment]
I should add that I have pored over the documentation to no avail.  I cannot find any reason for the failures.

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***

**Comment by abuzuhri** [comment]
Send me email with all details here and delete it from here not save to keep it .


I will try from my side and check when I’m in front of my laptop, but from first view it’s look like you need to do all steps again , I can help in that just send email and I will try to solve this with you

**Comment by booktrakker** [comment]
Not entirely clear what you are asking.  I am not following ‘here’.  Will be glad to comply, just need to know what the first sentence needs me to do.

 

Or are you saying that this is a public forum thus is not safe to post certain key codes and the like?  Got it.  That makes sense.  Do you have a means by which I can PM you?

 

I am also not super clear on what steps I would need to do again, but once I do I will be glad to comply.  If you can shed a bit more light on what steps you think I might need to review I will gladly do so.

 

If you are referring to the steps outlined in the Amazon docs, I have gone over all that repeatedly with my TAM, and we know it is set up properly as the test code can retrieve an order.  I cannot do the same thing using your library.

 

I can offer that they use very different authentication methods, and I am not at all clear as to what the differences are between them. All I know is that it works if I use their difficult SDK.

 

My apologies for being dense, but I am doing the best I can in a context where I am feeling a bit overwhelmed.

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***

**Comment by booktrakker** [comment]
Tareq,

I am not really sure what details you need - if you can give me some idea I will gladly provide them, but the above message has the two different approaches and should provide what you need.  I am at a loss as to what else you might need.

I have reviewed my IAM configuration with my Amazon TAM and they have confirmed that it is set up properly - which is confirmed by my success at using the Amazon SDK directly.  So the question is, why is there such a difference in how the Authorization is handled, and what do I need to do to make your approach work given that I can use the LWA Authentication method my TAM had me use?

It seems to be down to who can help me make this work.  Amazon will not help me with anything associated with using your library even if the issue is related to Amazon code - if I am not mistaken, the Amazon Credentials involved in your code is from Amazon.

For whatever reason I get the ARN error with your approach and I get an Order List if I use the code directly.  If we can identify the difference between how you are authorizing and the model they had me use - shown in the earlier message - then maybe I can get your library working.  I would very much prefer that as it seems much simpler, but I am running out of time and must do something now.

What puzzles me is why one  of the two methods uses the RoleARN and the other does not.  Why did you base the authorization workflow on the AmazonCredentials when the model they had me use takes such a different approach?  Can I use their approach with your library?

I have the deadlines looming and really have to move forward.  I am hoping you can spot whatever the detail is that is causing this.

I really appreciate your help!  Looking forward to seeing if there is a solution at hand.

**Comment by booktrakker** [comment]
I tried to run the Sample Code but it is complaining about missing a 'secrets.json' file.  I searched for such a file under the project folders and did not find any.

I am also not clear on what to do about the Role based ARN.  I configured a Role but it does not work, and the Amazon code does not use it, so I am stuck between Amazon code that I cannot make work and the really nice library you provide that I cannot make work.  All with Amazon imposing a deadline on us.  If I can figure out why my Role is not working perhaps I can get your library working.

I tried adding the STS policy to my Role ARN and it did not work.  How do I configure a Role for use with your Library?  Nothing I do seems to work, and my Amazon TAM will not touch this, so I am dead in the water at this point.  I cannot make either system work, and it is a matter of who manages to help me get it working first.  I am not thrilled with all this, but it is where I am at.

I can provide some screen shots of my IAM configuration if that would help, please direct me to what you would need to see and I will gladly provide the info.

**Comment by booktrakker** [comment]
Tareq,
I really could use your help as the Amazon folks are not responding and I cannot even build the Swagger models successfully.  Your library appears easy to use but due to differences in how you hand Authorization and Credentials I am unable to use your library.

One thing I am not clear on is why there is a difference in how Authorization is handled.  The model my Amazon TAM demonstrated and which worked is quite different from what you do, and after inspecting your code, I cannot see why or how.

Their method does not use a Role ARN, which is what is tripping me up with your approach.  They use two sets of credentials as shown above, and that works.

Can I convert your code to use that model?  Please help me understand what I need to do either to convert to the form used by my Amazon TAM or help me understand how to set up the Role.  I have reviewed the docs and procedures and cannot see any reason it should fail.

**Reply by abuzuhri** [reply]
> Tareq, I really could use your help as the Amazon folks are not responding and I cannot even build the Swagger models successfully. Your library appears easy to use but due to differences in how you hand Authorization and Credentials I am unable to use your library.
> 
> One thing I am not clear on is why there is a difference in how Authorization is handled. The model my Amazon TAM demonstrated and which worked is quite different from what you do, and after inspecting your code, I cannot see why or how.
> 
> Their method does not use a Role ARN, which is what is tripping me up with your approach. They use two sets of credentials as shown above, and that works.
> 
> Can I convert your code to use that model? Please help me understand what I need to do either to convert to the form used by my Amazon TAM or help me understand how to set up the Role. I have reviewed the docs and procedures and cannot see any reason it should fail.

Please message me in private email (tareq.abuzuhri@gmail.com) , I will help you solve this problem .

**Reply by bmarzano** [reply]
Was this ever resolved?  I'm having the same issue.  I can make the calls in Postman directly to Amazon just fine, but this library would make it all infinitely easier.  But I can't get past the "not authorized to perform: sts:AssumeRole" issue using this library.

**Comment by booktrakker** [comment]
I did not use this library due to concerns that Amazon would not provide support so I opted for the official Amazon SDK only to learn that they do not provide support for any SDK in terms of how they respond to issues......

**Comment by J-W-Chan** [comment]
This question has nothing to do with Amazon's official development kit. Because SP-API provides the open API standard, and anyone can generate their own SDK client based on it.
You need to ensure that your AWS user and SP-API developer account are normal. 
You can use postman to test if
![image](https://github.com/abuzuhri/Amazon-SP-API-CSharp/assets/6493028/f464f0c5-984d-456c-8717-855666c54d8b)

**Comment by booktrakker** [comment]
Hello,

 

I determined that I needed to update the two files from the SDK that handle the Signing.  Once I updated those two files it worked.

 

I did not get any meaningful assistance from Developer Support, I had to figure this out entirely on my own.  I am not impressed with this at all.  It caused a two week outage primarily due to my thinking that Support would actually try to help me, but due to using this SDK they basically washed their hands of it.

 

Whatever the reason, it lead to a two week outage and a loss of customers.

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***

**Comment by tildy** [comment]
Hello,

Do you mind sharing what you files you had to replace?  Been on this for
hours and still no luck.

**Comment by booktrakker** [comment]
Here are the two files I updated:

 

AWSSignerHelper.cs

AWSSigV4Signer.cs

 

Once I replaced them everything started working again.

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***

**Comment by tildy** [comment]
So I updated the 2 files from the SDK to the source project and now it will
not build.

**Comment by booktrakker** [comment]
This is where my contribution ends.  I updated them, it built, and solve my problem, so beyond this I cannot offer anything.

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***
